src/shogun/modelselection/NewGridSearch.cpp

```diff
@@ -0,0 +1,891 @@
+{
+	return attach(param, node.get());
+}
+
+ParameterNode*
+ParameterNode::attach(const std::string& param, ParameterNode* node)
+{
+	auto* node_copy = new ParameterNode(*node);
+	if (!set_param_helper(param, node_copy))
+		SG_SERROR("Could not attach %s\n", param.c_str());
+	return this;
+}
+
+ParameterNode*
+ParameterNode::attach(const std::string& param, ParameterNode&& node)
+{
+	if (!set_param_helper(param, &node))
+		SG_SERROR("Could not attach %s\n", param.c_str());
+	return this;
+}
+
+bool ParameterNode::set_param_helper(
+    const std::string& param, ParameterNode* node)
+{
+	auto param_iter = param.find(delimiter);
+
+	if (param_iter != std::string::npos)
+	{
+		return set_param_helper(
+		    param.substr(param_iter + delimiter.size()), node);
+	}
+	else
+	{
+		if (m_parent->has(param))
+		{
+			create_node(param, node);
+			return true;
+		}
+	}
+	return false;
+}
+
+std::string ParameterNode::to_string() const noexcept
+{
+	std::stringstream ss;
+	std::unique_ptr<AnyVisitor> visitor(new ToStringVisitor(&ss));
+	ss << get_name() << "(" << get_parent_name() << "(";
+
+	for (auto param_pair = m_param_mapping.begin();
+	     param_pair != m_param_mapping.end(); ++param_pair)
+	{
+		ss << param_pair->first.c_str() << "=";
+		param_pair->second.visit(visitor.get());
+		if (std::next(param_pair) != m_param_mapping.end())
+			ss << ", ";
+	}
+
+	if (!m_nodes.empty())
+	{
+		if (!m_param_mapping.empty())
+			ss << ", ";
+
+		for (auto node = m_nodes.begin(); node != m_nodes.end(); ++node)
+		{
+			ss << (*node).first << "=";
+			if (node->second.size() > 1)
+			{
+				ss << "[";
+			}
+			for (auto inner_node = node->second.begin();
+			     inner_node != node->second.end(); ++inner_node)
+			{
+				(*inner_node)->to_string_helper(ss, visitor);
+				if (std::next(inner_node) != node->second.end())
+					ss << ", ";
+			}
+			if (node->second.size() > 1)
+			{
+				ss << "]";
+			}
+			if (std::next(node) != m_nodes.end())
+				ss << ", ";
+		}
+	}
+	ss << "))";
+	return ss.str();
+}
+
+void ParameterNode::to_string_helper(
+    std::stringstream& ss, std::unique_ptr<AnyVisitor>& visitor) const noexcept
+{
+	ss << get_parent_name() << "(";
+	for (auto param_pair = m_param_mapping.begin();
+	     param_pair != m_param_mapping.end(); ++param_pair)
+	{
+		ss << param_pair->first.c_str() << "=";
+		param_pair->second.visit(visitor.get());
+		if (std::next(param_pair) != m_param_mapping.end())
+			ss << ", ";
+	}
+	ss << ")";
+}
+
+bool ParameterNode::set_param_helper(const std::string& param, const Any& value)
+{
+	auto param_iter = param.find(delimiter);
+	SG_SDEBUG("ParameterNode::set_param_helper - param: %s\n", param.c_str())
+
+	if (param_iter != std::string::npos)
+	{
+		SG_SDEBUG(
+		    "ParameterNode::set_param_helper - Looking for param in nodes: "
+		    "%s\n",
+		    param.c_str())
+		for (auto& node : m_nodes)
+		{
+			if (node.second.size() > 1)
+			{
+				SG_SERROR(
+				    "Ambiguous call! More than one %s found in "
```

Comment:
this is impossible or? Parameters cannot be registered twice...

Reply:
but you can have several kernels in a node, i.e. svm with gaussian and poly kernel

Reply:
ah yes or course! my bad!